### PR TITLE
feat(sink): add pulsar sink

### LINKED ISSUE #199: Support pulsar sink
## Describe the feature
A pulsar sink for Loggie.

### PR DESCRIPTION
**Pull request by carrotshub**
#### Proposed Changes:

* 增加支持pulsar sink
*
*

#### Which issue(s) this PR fixes:

Fixes #199

#### Additional documentation:

# pulsar

使用sink kafka将日志数据发送至下游Pulsar。

!!! example

    ```yaml
    sink:
      type: pulsar
      url: pulsar://localhost:6650
      topic: persistent://tenant/namespace/topic
    ```

## brokers

|    `字段`   |    `类型`    |  `是否必填`  |  `默认值`  |  `含义`  |
| ---------- | ----------- | ----------- | --------- | -------- |
| url| string  |    必填    |   无  | 日志发送端pulsar连接地址 |

## topic

|    `字段`   |    `类型`    |  `是否必填`  |  `默认值`  |  `含义`  |
| ---------- | ----------- | ----------- | --------- | -------- |
| topic | string  |    必填    |   无  | 发送日志至pulsar的topic |

## operation_timeout_seconds

|    `字段`   |    `类型`    |  `是否必填`  |  `默认值`  |  `含义`  |
| ---------- | ----------- | ----------- | --------- | -------- |
| operation_timeout_seconds| time.Duration|    非必填    |   30s  | Producer-create, subscribe and unsubscribe operations will be retried until this interval, after which the operation will be marked as failed |

## connectionTimeout

|    `字段`   |    `类型`    |  `是否必填`  |  `默认值`  |  `含义`  |
| ---------- | ----------- | ----------- | --------- | -------- |
| connectionTimeout| time.Duration|    非必填    |   5s | Timeout for the establishment of a TCP connection |

## sendTimeout

|    `字段`   |    `类型`    |  `是否必填`  |  `默认值`  |  `含义`  |
| ---------- | ----------- | ----------- | --------- | -------- |
| sendTimeout| time.Duration|    非必填    |   30s | SendTimeout set the timeout for a message that is not acknowledged by the server 30s |

## maxPendingMessages

|    `字段`   |    `类型`    |  `是否必填`  |  `默认值`  |  `含义`  |
| ---------- | ----------- | ----------- | --------- | -------- |
| sendTimeout| time.Duration|    非必填    |  无 | MaxPendingMessages specifies the max size of the queue holding the messages pending to receive an acknowledgment from the broker |

## hashingSchema

|    `字段`   |    `类型`    |  `是否必填`  |  `默认值`  |  `含义`  |
| ---------- | ----------- | ----------- | --------- | -------- |
| hashingSchema| int|    非必填    |  0 |HashingScheme is used to define the partition on where to publish a particular message. 0:JavaStringHash，1:Murmur3_32Hash |

## hashingSchema

|    `字段`   |    `类型`    |  `是否必填`  |  `默认值`  |  `含义`  |
| ---------- | ----------- | ----------- | --------- | -------- |
| hashingSchema| int|    非必填    |  0 |HashingScheme is used to define the partition on where to publish a particular message. 0:JavaStringHash，1:Murmur3_32Hash |

## compressionType

|    `字段`   |    `类型`    |  `是否必填`  |  `默认值`  |  `含义`  |
| ---------- | ----------- | ----------- | --------- | -------- |
| compressionType | int|    非必填    |   0 | 0:NoCompression,1:LZ4,2:ZLIB,3:ZSTD |

## LogLevel

|    `字段`   |    `类型`    |  `是否必填`  |  `默认值`  |  `含义`  |
| ---------- | ----------- | ----------- | --------- | -------- |
| LogLevel| string  |    非必填    |   0 | 日志级别: "info","debug", "error" |

## batchingMaxSize

|    `字段`   |    `类型`    |  `是否必填`  |  `默认值`  |  `含义`  |
| ---------- | ----------- | ----------- | --------- | -------- |
| batchingMaxSize| int  |    非必填    |   2048(KB)  | BatchingMaxSize specifies the maximum number of bytes permitted in a batch |

## batchingMaxMessages

|    `字段`   |    `类型`    |  `是否必填`  |  `默认值`  |  `含义`  |
| ---------- | ----------- | ----------- | --------- | -------- |
| batchingMaxMessages| int  |    非必填    |   1000  |BatchingMaxMessages specifies the maximum number of messages permitted in a batch |

## batchingMaxPublishDelay

|    `字段`   |    `类型`    |  `是否必填`  |  `默认值`  |  `含义`  |
| ---------- | ----------- | ----------- | --------- | -------- |
| batchingMaxPublishDelay| time.Duration  |    非必填    |   10ms  | BatchingMaxPublishDelay specifies the time period within which the messages sent will be batched |

## useTLS

|    `字段`   |    `类型`    |  `是否必填`  |  `默认值`  |  `含义`  |
| ---------- | ----------- | ----------- | --------- | -------- |
| useTLS| bool  |    非必填    |   false  | 是否使用TLS认证 |

## tlsTrustCertsFilePath

|    `字段`   |    `类型`    |  `是否必填`  |  `默认值`  |  `含义`  |
| ---------- | ----------- | ----------- | --------- | -------- |
| tlsTrustCertsFilePath| string  |    非必填    |   无  | the path to the trusted TLS certificate file |

## tlsAllowInsecureConnection

|    `字段`   |    `类型`    |  `是否必填`  |  `默认值`  |  `含义`  |
| ---------- | ----------- | ----------- | --------- | -------- |
| tlsAllowInsecureConnection| bool|    非必填    |   false  | Configure whether the Pulsar client accept untrusted TLS certificate from broker |

## certificatePath

|    `字段`   |    `类型`    |  `是否必填`  |  `默认值`  |  `含义`  |
| ---------- | ----------- | ----------- | --------- | -------- |
| certificatePath| string |    非必填    |   无 | TLS证书路径 |

## privateKeyPath

|    `字段`   |    `类型`    |  `是否必填`  |  `默认值`  |  `含义`  |
| ---------- | ----------- | ----------- | --------- | -------- |
| privateKeyPath| string |    非必填    |   无 | TLS privateKey路径 |

## token

|    `字段`   |    `类型`    |  `是否必填`  |  `默认值`  |  `含义`  |
| ---------- | ----------- | ----------- | --------- | -------- |
| token | string|    非必填    |  无  |  如果使用token认证鉴权pulsar，请填写此项|

## tokenFilePath

|    `字段`   |    `类型`    |  `是否必填`  |  `默认值`  |  `含义`  |
| ---------- | ----------- | ----------- | --------- | -------- |
| tokenFilePath| string|    非必填    |  无  |  auth token from a file|

### CODE REVIEW

**Comment by carrotshub** [comment]
@ethfoo

**Comment by carrotshub** [comment]
这是来自QQ邮箱的自动回复邮件。您好，您的邮件我已收到，我将尽快回复！谢谢！

File: pkg/sink/pulsar/config.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+		}
+	}
+	if len(config.Token) > 0 {
+		clientOptions.Authentication = pulsar.NewAuthenticationToken(string(config.Token))
+	}
+	if len(config.TokenFilePath) > 0 {
+		clientOptions.Authentication = pulsar.NewAuthenticationTokenFromFile(config.TokenFilePath)
+	}
+
+	standardLogger := logrus.StandardLogger()
```
Comment (mmaxiaolei) [inline pkg/sink/pulsar/config.go:100]: Can the log implementation use the [zerolog](https://github.com/rs/zerolog) used by loggie

**Comment by carrotshub** [comment]
> @carrotshub 麻烦冲突解决一下

已解决

File: pkg/sink/pulsar/config.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	CompressionType         pulsar.CompressionType `yaml:"compressionType,omitempty"`
+	Batching                bool                   `yaml:"batching,omitempty"`
+	BatchingMaxPublishDelay time.Duration          `yaml:"batchingMaxPublishDelay,omitempty"`
+	BatchingMaxMessages     uint                   `yaml:"batchingMaxMessages,omitempty"`
+	BatchingMaxSize         uint                   `yaml:"batchingMaxSize,omitempty"`
+}
+
+func (c *pulsarConfig) Validate() error {
+	if len(c.URL) == 0 {
+		return errors.New("no URL configured")
```
Comment (mmaxiaolei) [inline pkg/sink/pulsar/config.go:49]: use yaml hit: `validate:"required"`

File: pkg/sink/pulsar/config.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+	MaxPendingMessages      int                    `yaml:"maxPendingMessages,omitempty"`
+	HashingScheme           pulsar.HashingScheme   `yaml:"hashingSchema,omitempty"`
+	CompressionType         pulsar.CompressionType `yaml:"compressionType,omitempty"`
+	Batching                bool                   `yaml:"batching,omitempty"`
+	BatchingMaxPublishDelay time.Duration          `yaml:"batchingMaxPublishDelay,omitempty"`
+	BatchingMaxMessages     uint                   `yaml:"batchingMaxMessages,omitempty"`
+	BatchingMaxSize         uint                   `yaml:"batchingMaxSize,omitempty"`
+}
+
+func (c *pulsarConfig) Validate() error {
```
Comment (mmaxiaolei) [inline pkg/sink/pulsar/config.go:47]: `Validate()` can be replaced by yaml hit. such as:  
* `validate:"required,gte=0"`
* `validate:"required",dive`
*  and more: https://github.com/go-playground/validator/

File: pkg/sink/pulsar/config.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+	if c.CompressionType < 0 {
+		return errors.New("compression_type is incorrect")
+	}
+	return nil
+}
+
+func getOptions(
+	config *pulsarConfig,
+) (*pulsar.ClientOptions, *pulsar.ProducerOptions, error) {
+	err := config.Validate()
```
Comment (mmaxiaolei) [inline pkg/sink/pulsar/config.go:80]: When the pipeline start, `Validate()` will be called by loggie before the component `Init()`

File: pkg/sink/pulsar/config.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+		URL: config.URL,
+	}
+	if config.UseTLS {
+		clientOptions.TLSTrustCertsFilePath = config.TLSTrustCertsFilePath
+		if len(config.CertificatePath) > 0 {
+			clientOptions.Authentication = pulsar.NewAuthenticationTLS(config.CertificatePath, config.PrivateKeyPath)
+		}
+	}
+	if len(config.Token) > 0 {
+		clientOptions.Authentication = pulsar.NewAuthenticationToken(string(config.Token))
```
Comment (mmaxiaolei) [inline pkg/sink/pulsar/config.go:94]: the type of config.Token is string, why conversion?

File: pkg/sink/pulsar/sink.go [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+func (s *Sink) Init(context api.Context) error {
+	s.topicMatcher = pattern.MustInitMatcher(s.config.Topic)
+	return nil
+}
+
+func (s *Sink) Start() error {
+	c := s.config
+	// 初始化pulsar客户端
```
Comment (mmaxiaolei) [inline pkg/sink/pulsar/sink.go:70]: use english please

File: pkg/sink/pulsar/config.go [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+	Properties map[string]string `yaml:"properties,omitempty"`
+	// SendTimeout set the timeout for a message that is not acknowledged by the server 30s.default 30s
+	SendTimeout time.Duration `yaml:"sendTimeout,omitempty"`
+	// MaxPendingMessages set the max size of the queue holding the messages pending to receive an acknowledgment from the broker.
+	MaxPendingMessages      int                    `yaml:"maxPendingMessages,omitempty"`
+	HashingScheme           pulsar.HashingScheme   `yaml:"hashingSchema,omitempty"`
+	CompressionType         pulsar.CompressionType `yaml:"compressionType,omitempty"`
+	Batching                bool                   `yaml:"batching,omitempty"`
+	BatchingMaxPublishDelay time.Duration          `yaml:"batchingMaxPublishDelay,omitempty"`
+	BatchingMaxMessages     uint                   `yaml:"batchingMaxMessages,omitempty"`
```
Comment (mmaxiaolei) [inline pkg/sink/pulsar/config.go:43]: Think about default 2048 as loggie batch size

File: pkg/sink/pulsar/sink.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+	var sendErr error
+	for i := range events {
+		event := &events[i]
+		serializerEncode, err := s.cod.Encode(*event)
+		if err != nil {
+			log.Warn("encode event error: %+v", err)
+			return result.Fail(err)
+		}
+		pTime := time.Now()
+		s.clt.producer.SendAsync(context.Background(), &pulsar.ProducerMessage{
```
Comment (mmaxiaolei) [inline pkg/sink/pulsar/sink.go:126]: SendAsync will lost data?

File: pkg/sink/pulsar/config.go (hunk 1 above)
Comment (carrotshub) [reply pkg/sink/pulsar/config.go:100]: zerolog does not implement some methods of the pulsar SDK logger interface

File: pkg/sink/pulsar/config.go (hunk 2 above)
Comment (carrotshub) [reply pkg/sink/pulsar/config.go:49]: Agree, I will modify

File: pkg/sink/pulsar/config.go (hunk 3 above)
Comment (carrotshub) [reply pkg/sink/pulsar/config.go:47]: Agree, I will modify

File: pkg/sink/pulsar/config.go (hunk 4 above)
Comment (carrotshub) [reply pkg/sink/pulsar/config.go:80]: I will modify it according to your suggestion

File: pkg/sink/pulsar/config.go (hunk 5 above)
Comment (carrotshub) [reply pkg/sink/pulsar/config.go:94]: sorry，sorry, I forgot to delete the code left in the development verification stage

File: pkg/sink/pulsar/sink.go (hunk 6 above)
Comment (carrotshub) [reply pkg/sink/pulsar/sink.go:70]: OK

File: pkg/sink/pulsar/sink.go [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+			EventTime: pTime,
+			Key:       fmt.Sprintf("%d", pTime.Nanosecond()),
```
[the remark this answers is not in the document]
Comment (carrotshub) [reply pkg/sink/pulsar/sink.go:128]: Messages are optionally tagged with keys, which is useful for things like [topic compaction](https://pulsar.apache.org/docs/concepts-topic-compaction).
But I do not have any idea to set or support the configration from header

File: pkg/sink/pulsar/config.go (hunk 7 above)
Comment (carrotshub) [reply pkg/sink/pulsar/config.go:43]: ok, this will be set

File: pkg/sink/pulsar/sink.go (hunk 8 above)
Comment (carrotshub) [reply pkg/sink/pulsar/sink.go:126]: In theory, no data will be lost，Send a message, this call will be blocking until is successfully acknowledged by the Pulsar broker.

**Comment by ethfoo** [comment]
resovle confilcts by: https://github.com/loggie-io/loggie/pull/417

**Comment by carrotshub** [comment]
这是来自QQ邮箱的自动回复邮件。您好，您的邮件我已收到，我将尽快回复！谢谢！

**Pull request closed without merging**